crates/shim/src/asynchronous/mod.rs

```diff
@@ -0,0 +1,429 @@
+///
+/// Start and delete routines will be called to handle containerd's shim lifecycle requests.
+#[async_trait]
+pub trait Shim {
+    /// Type to provide task service for the shim.
+    type T: Task + Send + Sync;
+
+    /// Create a new instance of  async Shim.
+    ///
+    /// # Arguments
+    /// - `runtime_id`: identifier of the container runtime.
+    /// - `id`: identifier of the shim/container, passed in from Containerd.
+    /// - `namespace`: namespace of the shim/container, passed in from Containerd.
+    /// - `publisher`: publisher to send events to Containerd.
+    /// - `config`: for the shim to pass back configuration information
+    async fn new(
+        runtime_id: &str,
+        id: &str,
+        namespace: &str,
+        publisher: RemotePublisher,
+        config: &mut Config,
+    ) -> Self;
+
+    /// Start shim will be called by containerd when launching new shim instance.
+    ///
+    /// It expected to return TTRPC address containerd daemon can use to communicate with
+    /// the given shim instance.
+    /// See https://github.com/containerd/containerd/tree/master/runtime/v2#start
+    /// this is an asynchronous call
+    async fn start_shim(&mut self, opts: StartOpts) -> Result<String>;
+
+    /// Delete shim will be called by containerd after shim shutdown to cleanup any leftovers.
+    /// this is an asynchronous call
+    async fn delete_shim(&mut self) -> Result<DeleteResponse>;
+
+    /// Wait for the shim to exit asynchronously.
+    async fn wait(&mut self);
+
+    /// Create the task service object asynchronously.
+    async fn create_task_service(&self) -> Self::T;
+}
+
+/// Async Shim entry point that must be invoked from tokio `main`.
+pub async fn run<T>(runtime_id: &str, opts: Option<Config>)
+where
+    T: Shim + Send + Sync + 'static,
+{
+    if let Some(err) = bootstrap::<T>(runtime_id, opts).await.err() {
+        eprintln!("{}: {:?}", runtime_id, err);
+        let mut f = tokio::fs::OpenOptions::new()
+            .create(true)
+            .write(true)
+            .append(true)
+            .open("/tmp/containerd-shim.log")
+            .await
+            .unwrap();
+        f.write_all(format!("error bootstrap {}....\n", err).as_bytes())
+            .await
+            .unwrap();
+        process::exit(1);
+    }
+}
+
+async fn bootstrap<T>(runtime_id: &str, opts: Option<Config>) -> Result<()>
+where
+    T: Shim + Send + Sync + 'static,
+{
+    // Parse command line
+    let os_args: Vec<_> = env::args_os().collect();
+    let flags = args::parse(&os_args[1..])?;
+
+    let ttrpc_address = env::var(TTRPC_ADDRESS)?;
+    let publisher = publisher::RemotePublisher::new(&ttrpc_address).await?;
+
+    // Create shim instance
+    let mut config = opts.unwrap_or_else(Config::default);
+
+    // Setup signals
+    let signals = setup_signals_tokio(&config);
+
+    if !config.no_sub_reaper {
+        asyncify(|| -> Result<()> { reap::set_subreaper() }).await?;
```

Comment:
Why not just `reap::set_subreaper()`? I'd use async only where its applicable.

Reply:
I  thought the set_subreaper may be a blocking operation.  Maybe it is not? anyway, I followed your instruction to changed it to call `reap::set_subreaper()` directly